crates/lang_srv/src/registry.rs

```diff
@@ -1,75 +1,191 @@
                 }
             }
-            DocumentChange::Closed(_url) => {
-                // Do nothing.
+            None => {
+                documents.insert(url.clone(), DocumentPair::new(document.clone(), document));
+            }
+        }
+    }
+
+    pub async fn apply_changes<'a>(&self, analysed_docs: Vec<AnalyzedDocument>, updating_url: Url) {
+        let mut documents = self.documents.lock().await;
+        debug!(
+            "finised doc analysis for doc: {:?}",
+            updating_url.to_string()
+        );
+
+        for document in analysed_docs {
+            let document = Arc::new(document);
+            //Write the newly analysed document into the partial document that any request requiring the latest document will be waiting on
+            if document.doc_info.url == updating_url {
+                documents
+                    .get_mut(&updating_url)
+                    .map(|a| a.latest_document.set(document.clone()).unwrap());
             }
+            Registry::update_document(&mut documents, document);
         }
     }
 
-    fn document_by_url(&mut self, url: &Url) -> Option<&mut AnalyzedDocument> {
-        self.documents.get_mut(url)
+    pub async fn apply_doc_info_changes(&self, url: Url, info: DocInfo) {
+        let mut documents_lock = self.documents.lock().await;
+        let doc = documents_lock.get_mut(&url);
+        match doc {
+            Some(a) => {
+                debug!(
+                    "set the docInfo for {:?} to version:{:?}",
+                    url.as_str(),
+                    info.version
+                );
+                *a = DocumentPair {
+                    info,
+                    last_good_document: a.last_good_document.clone(),
+                    latest_document: OnceLock::new(),
+                };
+            }
+            None => debug!("no existing docinfo for {:?} ", url.as_str()),
+        }
     }
 
-    pub fn diagnostics(&mut self, url: &Url) -> Vec<Diagnostic> {
-        let Some(document) = self.document_by_url(url) else {
+    async fn document_info_by_url(&self, url: &Url) -> Option<DocInfo> {
+        self.documents.lock().await.get(url).map(|a| a.info.clone())
+    }
+
+    ///Tries to get the latest document from analysis.
+    ///Gives up and returns none after 5 seconds.
+    async fn latest_document_by_url(&self, url: &Url) -> Option<Arc<AnalyzedDocument>> {
+        let start = std::time::Instant::now();
+        let duration = std::time::Duration::from_secs(5);
+
+        while start.elapsed() < duration {
+            match self.documents.lock().await.get(url) {
+                Some(a) => match a.latest_document.get() {
+                    Some(a) => return Some(a.clone()),
+                    None => (),
+                },
+
+                None => return None,
+            }
+        }
+        warn!("Timed out trying to get latest document");
+        None
```

Comment:
This loop is hot and will contest heavy. I would restructure it to something like:


```
async fn wait_for_completion(&self, url: &Url): Arc<AnalyzedDocument> {
    loop {
       let docs = self.documents.lock().await;
       if let Some(a) = a.latest_document.get(url) {
          if let Some(a) = a.latest_document.get() { return a.clone(); }
       }
        drop(docs);
        tokio::task::yield_now().await;
    }
}

async fn latest_document_by_url(&self, url: &Url) -> Option<Arc<AnalyzedDocument>> {
  tokio::select! {
    _ = tokio::time::sleep(duration) => { None },
    doc = self.wait_for_completion(url) => { Some(doc) }
  }
}
```

Reply:
I completely forgot, I had meant to add a little delay after each iteration. I've adjusted it a little and added a 10ms sleep so we don't check stupidly often if nothing else is happening